{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 1\n",
    "Jupyter notebook corresponding to the input data for the phylogenetic tree with annotated typing schemes and plasmid presence/absence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import colorcet as cc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "plasmid_file = \"../0_Data/0_Raw/best_hits_1000bp_v11.csv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isolate</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>cp26</th>\n",
       "      <th>cp32-1</th>\n",
       "      <th>cp32-1+5</th>\n",
       "      <th>cp32-10</th>\n",
       "      <th>cp32-11</th>\n",
       "      <th>cp32-12</th>\n",
       "      <th>cp32-13</th>\n",
       "      <th>cp32-2</th>\n",
       "      <th>...</th>\n",
       "      <th>lp28-6</th>\n",
       "      <th>lp28-7</th>\n",
       "      <th>lp28-8</th>\n",
       "      <th>lp28-9</th>\n",
       "      <th>lp32-3</th>\n",
       "      <th>lp36</th>\n",
       "      <th>lp38</th>\n",
       "      <th>lp5</th>\n",
       "      <th>lp54</th>\n",
       "      <th>lp56</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASM1913465v1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASM215146v1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASM215148v1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASM215150v1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASM2466215v1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Isolate  chromosome  cp26  cp32-1  cp32-1+5  cp32-10  cp32-11  \\\n",
       "0  ASM1913465v1         1.0   1.0     0.0       1.0      0.0      0.0   \n",
       "1   ASM215146v1         1.0   1.0     1.0       1.0      0.0      0.0   \n",
       "2   ASM215148v1         1.0   1.0     0.0       1.0      0.0      0.0   \n",
       "3   ASM215150v1         1.0   1.0     0.0       1.0      0.0      0.0   \n",
       "4  ASM2466215v1         1.0   1.0     0.0       0.0      1.0      1.0   \n",
       "\n",
       "   cp32-12  cp32-13  cp32-2  ...  lp28-6  lp28-7  lp28-8  lp28-9  lp32-3  \\\n",
       "0      0.0      0.0     0.0  ...     0.0     0.0     1.0     0.0     0.0   \n",
       "1      0.0      0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0   \n",
       "2      0.0      0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0   \n",
       "3      0.0      0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0   \n",
       "4      0.0      0.0     0.0  ...     1.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "   lp36  lp38  lp5  lp54  lp56  \n",
       "0   1.0   0.0  0.0   1.0   1.0  \n",
       "1   1.0   1.0  0.0   1.0   1.0  \n",
       "2   1.0   0.0  0.0   1.0   1.0  \n",
       "3   1.0   1.0  0.0   1.0   1.0  \n",
       "4   1.0   1.0  0.0   1.0   1.0  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plasmids = pd.read_csv(plasmid_file)\n",
    "plasmids = plasmids.rename(columns={\"assembly_id\": \"Isolate\"})\n",
    "\n",
    "# clean isolate names\n",
    "plasmids['Isolate'] = plasmids['Isolate'].str.replace(r'GCF_\\d+\\.\\d+_', '', regex=True)\n",
    "plasmids['Isolate'] = plasmids['Isolate'].str.replace(r'_genomic', '', regex=True)\n",
    "plasmids = plasmids[[\"Isolate\", \"plasmid_name\"]].drop_duplicates().copy()\n",
    "\n",
    "# convert plasmid annotations to a presence/absence matrix\n",
    "plasmids = (plasmids.assign(value=1)\n",
    "             .pivot(index='Isolate', columns='plasmid_name', values='value')\n",
    "             .fillna(0)\n",
    "             .reset_index()\n",
    "             .rename_axis(None, axis=1)\n",
    ")\n",
    "plasmids.iloc[:, 1:] = plasmids.iloc[:, 1:].astype(int)\n",
    "\n",
    "plasmids.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ASM1913465v1', 'ASM215146v1', 'ASM215148v1', 'ASM215150v1',\n",
       "       'ASM2466215v1', 'ASM2466217v1', 'ASM2466219v1', 'ASM336729v1',\n",
       "       'ASM4079071v1', 'ASM4079073v1', 'ASM4079074v1', 'ASM4079075v1',\n",
       "       'ASM4079076v1', 'ASM4079077v1', 'ASM4079078v1', 'ASM4079079v1',\n",
       "       'ASM4079080v1', 'B331P', 'B418P', 'B500P', 'ESI26H', 'ESI361H',\n",
       "       'ESI403H', 'ESI425H', 'PFhe_I_PB_Ill_cons', 'UCT109H', 'UCT110H',\n",
       "       'UCT113H', 'UCT124H', 'UCT29H', 'UCT30H', 'UCT31H', 'UCT32H',\n",
       "       'UCT35H', 'UCT50H', 'UCT92H', 'UCT96H', 'UNY1032P', 'UNY1038P',\n",
       "       'UNY1083P', 'UNY1085P', 'UNY1090P', 'UNY1128P', 'UNY149P',\n",
       "       'UNY169P', 'UNY172P', 'UNY193P', 'UNY203P', 'UNY208P', 'UNY990P',\n",
       "       'URI101H', 'URI102H', 'URI103H', 'URI107H', 'URI111H', 'URI112H',\n",
       "       'URI117H', 'URI118H', 'URI120H', 'URI33H', 'URI34H', 'URI36H',\n",
       "       'URI39H', 'URI40H', 'URI41H', 'URI42H', 'URI44H', 'URI46H',\n",
       "       'URI47H', 'URI48H', 'URI56H', 'URI86H', 'URI87H', 'URI88H',\n",
       "       'URI89H', 'URI91H', 'URI93H', 'UWI247P', 'UWI248P', 'UWI263P',\n",
       "       'UWI283P', 'XYZ459H'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check isolates in sample\n",
    "plasmids[\"Isolate\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "plasmids = plasmids.drop(columns=[\"chromosome\"])\n",
    "\n",
    "# conversion to integers\n",
    "for i in plasmids.columns.drop(\"Isolate\"):\n",
    "    plasmids[i] = plasmids[i].apply(lambda z: 0 if z==0 else 1)\n",
    "\n",
    "# sorting the plasmids by overall prevalence in the dataset    \n",
    "column_sums = plasmids[plasmids.columns.drop(\"Isolate\")].sum()\n",
    "sorted_columns = column_sums.sort_values(ascending=False)\n",
    "plasmids_rearranged = plasmids[sorted_columns.index].copy(deep=True)\n",
    "plasmids = pd.concat([plasmids[\"Isolate\"], plasmids_rearranged], axis=1)\n",
    "\n",
    "plasmid_list = plasmids.columns.drop(\"Isolate\")\n",
    "\n",
    "# data processing\n",
    "for i in plasmid_list:\n",
    "    plasmids[i] = plasmids[i].apply(lambda z: \"Absent\" if z==0 else i)\n",
    "\n",
    "ordered_plasmids = list(plasmids.columns.drop(\"Isolate\"))\n",
    "ordered_plasmids.append(\"Unclassified\")\n",
    "ordered_plasmids.append(\"chromosome\")\n",
    "\n",
    "# producing a plasmid-specific color mapping\n",
    "palette = sns.color_palette(cc.glasbey_light, n_colors=len(ordered_plasmids)).as_hex()\n",
    "plasmid_color_scheme = dict(zip(ordered_plasmids, palette))\n",
    "with open('../0_Data/1_Intermediate/plasmid_color_scheme.pkl', 'wb') as pickle_file:\n",
    "    pickle.dump(plasmid_color_scheme, pickle_file)\n",
    "\n",
    "# reshaping the data in preparation for tree-building\n",
    "plasmids_long = pd.melt(plasmids, id_vars=['Isolate'], var_name='plasmid', value_name='presence')\n",
    "plasmids_long[\"color\"] = plasmids_long[\"plasmid\"].apply(lambda z: plasmid_color_scheme[z])\n",
    "plasmids_long = plasmids_long[plasmids_long[\"presence\"]!=\"Absent\"].copy().reset_index(drop=True)\n",
    "plasmids_long.to_csv(\"../0_Data/2_Processed/plasmids_binary_long.csv\",\n",
    "                index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plasmid Color Scheme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cp26': '#d60000',\n",
       " 'lp54': '#018700',\n",
       " 'lp17': '#b500ff',\n",
       " 'lp36': '#05acc6',\n",
       " 'lp28-3': '#97ff00',\n",
       " 'lp25': '#ffa52f',\n",
       " 'cp32-7': '#ff8ec8',\n",
       " 'lp28-4': '#79525e',\n",
       " 'lp38': '#00fdcf',\n",
       " 'cp32-9': '#afa5ff',\n",
       " 'cp32-6': '#93ac83',\n",
       " 'cp32-11': '#9a6900',\n",
       " 'lp28-1': '#366962',\n",
       " 'cp32-3': '#d3008c',\n",
       " 'cp32-1+5': '#fdf490',\n",
       " 'cp32-4': '#c86e66',\n",
       " 'cp32-12': '#9ee2ff',\n",
       " 'lp56': '#00c846',\n",
       " 'lp28-6': '#a877ac',\n",
       " 'cp32-8': '#b8ba01',\n",
       " 'lp28-2': '#f4bfb1',\n",
       " 'cp32-1': '#ff28fd',\n",
       " 'cp32-2': '#f2cdff',\n",
       " 'lp28-5': '#009e7c',\n",
       " 'cp32-5': '#ff6200',\n",
       " 'cp9': '#56642a',\n",
       " 'cp32-10': '#953f1f',\n",
       " 'lp21-cp9': '#90318e',\n",
       " 'lp28-7': '#ff3464',\n",
       " 'lp28-8': '#a0e491',\n",
       " 'lp21': '#8c9ab1',\n",
       " 'lp5': '#829026',\n",
       " 'cp32-13': '#ae083f',\n",
       " 'cp9-3': '#77c6ba',\n",
       " 'cp32-9-4': '#bc9157',\n",
       " 'lp28-9': '#e48eff',\n",
       " 'cp32-3+10': '#72b8ff',\n",
       " 'lp28-11': '#c6a5c1',\n",
       " 'lp32-3': '#ff9070',\n",
       " 'Unclassified': '#d3c37c',\n",
       " 'chromosome': '#bceddb'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dictionary assigning hex-codes to replicons\n",
    "plasmid_color_scheme"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plasmid Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Isolate\n",
       "URI40H          23\n",
       "UCT110H         22\n",
       "URI120H         22\n",
       "UCT124H         21\n",
       "UCT30H          21\n",
       "                ..\n",
       "ESI425H         11\n",
       "ASM215146v1     11\n",
       "ESI361H         10\n",
       "ASM215148v1     10\n",
       "ASM1913465v1    10\n",
       "Name: count, Length: 82, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count the number of plasmids per isolate\n",
    "plasmids_long[\"Isolate\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "plasmid\n",
       "cp26         1.000000\n",
       "lp54         1.000000\n",
       "lp17         0.975610\n",
       "lp36         0.914634\n",
       "lp28-3       0.865854\n",
       "lp25         0.853659\n",
       "cp32-7       0.829268\n",
       "lp28-4       0.804878\n",
       "lp38         0.756098\n",
       "cp32-9       0.719512\n",
       "cp32-6       0.695122\n",
       "cp32-11      0.670732\n",
       "lp28-1       0.670732\n",
       "cp32-3       0.634146\n",
       "cp32-1+5     0.621951\n",
       "cp32-4       0.475610\n",
       "cp32-12      0.463415\n",
       "lp56         0.451220\n",
       "lp28-6       0.451220\n",
       "cp32-8       0.439024\n",
       "lp28-2       0.426829\n",
       "cp32-1       0.329268\n",
       "cp32-2       0.329268\n",
       "lp28-5       0.317073\n",
       "cp32-5       0.292683\n",
       "cp9          0.268293\n",
       "cp32-10      0.231707\n",
       "lp21-cp9     0.182927\n",
       "lp28-8       0.085366\n",
       "lp28-7       0.085366\n",
       "lp21         0.073171\n",
       "lp5          0.073171\n",
       "cp32-13      0.073171\n",
       "cp9-3        0.048780\n",
       "cp32-9-4     0.036585\n",
       "lp28-9       0.024390\n",
       "cp32-3+10    0.024390\n",
       "lp28-11      0.012195\n",
       "lp32-3       0.012195\n",
       "Name: count, dtype: float64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count the prevalence of each plasmid across the sample\n",
    "plasmids_long[\"plasmid\"].value_counts()/len(plasmids_long[\"Isolate\"].unique())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "data_analysis_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
